dhis-2/dhis-services/dhis-service-program-rule/src/main/java/org/hisp/dhis/programrule/engine/RuleActionSendMessageImplementer.java

```diff
@@ -80,68 +64,34 @@ public boolean accept(RuleAction ruleAction) {
 
   @Override
   public void implement(RuleEffect ruleEffect, Enrollment enrollment) {
-    NotificationValidationResult result = validate(ruleEffect, enrollment);
-
-    if (!result.isValid()) {
-      return;
-    }
+    ProgramNotificationTemplate template =
+        notificationHelper.getNotificationTemplate(ruleEffect.getRuleAction());
 
-    ProgramNotificationTemplate template = result.getTemplate();
+    NotificationValidationResult result = notificationHelper.validate(template, enrollment);
 
-    String key = generateKey(template, enrollment);
+    if (result.isValid()) {
+      programNotificationService.sendProgramRuleTriggeredNotifications(template, enrollment);
 
-    publisher.publishEvent(new ProgramRuleEnrollmentEvent(this, template.getId(), enrollment));
-
-    if (result.getLogEntry() != null) {
-      return;
+      if (result.needsToCreateLogEntry()) {
+        notificationHelper.createLogEntry(template, enrollment);
+      }
     }
-
-    ExternalNotificationLogEntry entry = createLogEntry(key, template.getUid());
-    entry.setNotificationTriggeredBy(NotificationTriggerEvent.PROGRAM);
-    entry.setAllowMultiple(template.isSendRepeatable());
-
-    notificationLoggingService.save(entry);
   }
 
   @Override
   public void implement(RuleEffect ruleEffect, Event event) {
-    checkNotNull(event, "Event cannot be null");
+    ProgramNotificationTemplate template =
+        notificationHelper.getNotificationTemplate(ruleEffect.getRuleAction());
+    NotificationValidationResult result =
+        notificationHelper.validate(template, event.getEnrollment());
```

Comment:
will it work for programs without registration ?

Reply:
Yes.
In the validate method this is covered checking the enrollment and the program of the enrollment.